src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+                    {
+                        solution = solution.RemoveProjectReference(Id, projectReference);
+                    }
+
+                    ClearAndZeroCapacity(_projectReferencesRemovedInBatch);
+
+                    // Analyzer reference adding...
+                    solution = solution.AddAnalyzerReferences(Id, _analyzersAddedInBatch.Select(a => a.GetReference()));
+                    ClearAndZeroCapacity(_analyzersAddedInBatch);
+
+                    // Analyzer reference removing...
+                    foreach (var analyzerReference in _analyzersRemovedInBatch)
+                    {
+                        solution = solution.RemoveAnalyzerReference(Id, analyzerReference.GetReference());
+                    }
+
+                    ClearAndZeroCapacity(_analyzersRemovedInBatch);
+
+                    // Other property modifications...
+                    foreach (var propertyModification in _projectPropertyModificationsInBatch)
+                    {
+                        solution = propertyModification(solution);
+                    }
+
+                    ClearAndZeroCapacity(_projectPropertyModificationsInBatch);
+
+                    return solution;
+                });
+
+                foreach (var (documentId, textContainer) in documentsToOpen)
+                {
+                    _workspace.ApplyChangeToWorkspace(w => w.OnDocumentOpened(documentId, textContainer));
+                }
+
+                foreach (var (documentId, textLoader) in documentsToClose)
+                {
+                    _workspace.ApplyChangeToWorkspace(w => w.OnDocumentClosed(documentId, textLoader));
+                }
+
+                // Check for those files being opened to start wire-up if necessary
+                _workspace.CheckForOpenDocuments(documentFileNamesAdded.ToImmutable());
+            }
+        }
+
+        #endregion
+
+        #region Source File Addition/Removal
+
+        public void AddSourceFile(string fullPath, SourceCodeKind sourceCodeKind = SourceCodeKind.Regular, ImmutableArray<string> folders = default)
+        {
+            _sourceFiles.AddFile(fullPath, sourceCodeKind, folders);
+        }
+
+        public DocumentId AddSourceTextContainer(SourceTextContainer textContainer, string fullPath, SourceCodeKind sourceCodeKind = SourceCodeKind.Regular, ImmutableArray<string> folders = default)
+        {
+            return _sourceFiles.AddTextContainer(textContainer, fullPath, sourceCodeKind, folders);
+        }
+
+        public bool ContainsSourceFile(string fullPath)
+        {
+            return _sourceFiles.ContainsFile(fullPath);
+        }
+
+        public void RemoveSourceFile(string fullPath)
+        {
+            _sourceFiles.RemoveFile(fullPath);
+        }
+
+        public void RemoveSourceTextContainer(SourceTextContainer textContainer)
+        {
+            _sourceFiles.RemoveTextContainer(textContainer);
+        }
```

Comment:
Are these updates always batched? They seem to be processed only when batch scope is closed. What happens with these if there is no batch scope open?

Reply:
No; what might throw you off here is _sourceFiles is a type that is doing all the batching-or-not logic inside RemoveTextContainer; it's not a dumb list.

The logic for regular documents and additional documents is identical except for the methods it has to call to interact with the workspace, so it's refactored into a separate class and this is just forwarding to it.

Reply:
I see. I'd perhaps call the class `BatchingDocumentCollection` or something like that

---
In reply to: [215407300](https://github.com/dotnet/roslyn/pull/26931#discussion_r215407300) [](ancestors = 215407300)